finagle-base-http/src/main/scala/com/twitter/finagle/http/headers/JTreeMapBackedHeaderMap.scala

```diff
@@ -2,29 +2,102 @@ package com.twitter.finagle.http.headers
-    new java.util.TreeMap[String, Header.Root](JTreeMapBackedHeaderMap.SharedComparitor)
-
-    def getAll(key: String): Seq[String] = underlying.synchronized {
-      underlying.get(key.toLowerCase) match {
-        case null => Nil
-        case r: Header.Root => r.values
-      }
+final private class JTreeMapBackedHeaderMap extends HeaderMap {
+  import HeaderMap._
+
+  // In general, Map's that are not thread safe are not
+  // durable to concurrent modification and can result in infinite loops
+  // and exceptions.
+  // As such, we synchronize on the underlying collection when performing
+  // accesses to avoid this. In the common case of no concurrent access,
+  // this should be cheap.
+  val underlying: java.util.TreeMap[String, Header.Root] = 
+  new java.util.TreeMap[String, Header.Root](JTreeMapBackedHeaderMap.SharedComparitor)
+
+  private def foreachConsumer[U](f: ((String, String)) => U):
+    BiConsumer[String, Header.Root] = new BiConsumer[String, Header.Root](){
+      def accept(key: String, header: Header.Root): Unit = header.iterator.foreach(
+        nv => f(nv.name, nv.value)
+      )
+    }
+
+  override def foreach[U](f: ((String, String)) => U): Unit = 
+    underlying.forEach(foreachConsumer(f))
+
+  // ---- HeaderMap -----
+
+  // Validates key and value.
+  def add(key: String, value: String): this.type = {
+    validateName(key)
+    addUnsafe(key, foldReplacingValidateValue(key, value))
+  }
+
+  // Validates key and value.
+  def set(key: String, value: String): this.type = {
+    validateName(key)
+    setUnsafe(key, foldReplacingValidateValue(key, value))
+  }
+
+  // ---- Map/MapLike -----
+
+  def -=(key: String): this.type = removed(key)
+  def +=(kv: (String, String)): this.type = set(kv._1, kv._2)
+
+  /**
+   * Underlying headers eagerly copied to an array, without synchronizing
+   * on the underlying collection.
+   */
+  private[this] def copyHeaders: Iterator[Header.Root] = underlying.values.toArray(new Array[Header.Root](underlying.size)).iterator
+
+  def iterator: Iterator[(String, String)] =
+    nameValueIterator.map(nv => (nv.name, nv.value))
+
+  override def keys: Set[String] = keysIterator.toSet
+
+  override def keysIterator: Iterator[String] = underlying.synchronized {
+    //the common case has a single element in Headers. Prevent unneeded List
+    //allocations for that case (don't flatMap)
+    val valuesIterator = copyHeaders
+    var currentEntries: Iterator[String] = Iterator.empty
+    new AbstractIterator[String]{
+      def hasNext: Boolean = currentEntries.hasNext || valuesIterator.hasNext
+      def next(): String =
+        if (currentEntries.hasNext) currentEntries.next()
+        else {
+          val h = valuesIterator.next()
+          if (h.next == null) h.name
+          else {
+            currentEntries = h.iterator.map(nv => nv.name).toList.distinct.iterator
+            currentEntries.next()
+          }
+        }
+    }
+  }
```

Comment:
This looks pretty similar in result to `HeadersHash.uniqueNamesIterator`. Could we share that logic?

Reply:
I factored the (lazier) logic of HeaderHash out, and made both call that.